README.md
![CI](https://github.com/RoboticExplorationLab/ALTRO.jl/workflows/CI/badge.svg)
[![codecov](https://codecov.io/gh/RoboticExplorationLab/ALTRO.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/RoboticExplorationLab/ALTRO.jl)
[![](https://img.shields.io/badge/docs-dev-blue.svg)](https://roboticexplorationlab.org/Altro.jl/dev/)

# Altro.jl
Implementation of the Augmented Lagrangian TRajectory Optimizer (ALTRO) solver, a very fast solver for constrained trajectory optimization problems.
ALTRO uses iterative LQR (iLQR) with an augmented Lagrangian framework and can solve problems with nonlinear inequality and equality path constraints
and nonlinear dynamics. The key features of the ALTRO solver are:
  * General nonlinear cost functions, including minimum time problems
  * General nonlinear state and input constraints
  * Infeasible state initialization
  * Square-root methods for improved numerical conditioning
  * Active-set projection method for solution polishing

Altro.jl solves trajectory optimization problems set up using [TrajectoryOptimization.jl](https://github.com/RoboticExplorationLab/TrajectoryOptimization.jl).

For details on the solver, see the original [conference paper](http://roboticexplorationlab.org/papers/altro-iros.pdf) or related
[tutorial](https://bjack205.github.io/papers/AL_iLQR_Tutorial.pdf).

## Simple Example
See [examples/quickstart.jl](https://github.com/RoboticExplorationLab/Altro.jl/tree/master/examples/quickstart.jl) 
for a simple example of solving the canonical cartpole problem with Altro.jl.

## Solver Statistics
ALTRO logs intermediate values during the course of the solve. These values are all
stored in the `SolverStats` type, accessible via `solver.stats` or `Altro.stats(solver)`. This currently stores the following information:
| Field | Description |
| ----- | ----------- |
| `iterations` | Total number of iterations |
| `iterations_outer` | Number of outer loop (Augmented Lagrangian) iterations |
| `iterations_pn` | Number of projected newton iterations |
| `iteration` | Vector of iteration number |
| `iteration_outer` | Vector of outer loop iteration number |
| `cost` | Vector of costs |
| `dJ` | Change in cost |
| `c_max` | Maximum constrained violation |
| `gradient` | Approximation of dual optimality residual (2-norm of gradient of the Lagrangian) |
| `penalty_max` | Maximum penalty parameter |

The other fields are used interally by the solver and not important to the end user.

The vector fields of the `SolverStats` type can be converted to a dictionary via `Dict(stats::SolverStats)`,
which can then be cast into a tabular format such as `DataFrame` from DataFrames.jl.

## Solver Options
Like any nonlinear programming solver, ALTRO comes with a host of solver options.
While the default values yield good/acceptable performance on many problem, extra
performance can always be gained by tuning these parameters. In practice, there are
only a few parameters that need to be tuned. See the [AL-iLQR Tutorial](https://bjack205.github.io/papers/AL_iLQR_Tutorial.pdf) for more details.

The ALTRO solver is actually a composition of several different solvers with their own
options. Early versions of Altro.jl required the user to manipulate a rather confusing
heirarchy of solver options. Newer version of Altro.jl provide a single options struct
that dramatically simplifies setting and working with the solver parameters.

### Setting Solver Options
Solver options can be specified when the solver is instantiated or afterwards using 
the `set_options!` command. If we have a previously constructed `Problem`, this looks
like
```julia
solver = ALTROSolver(prob, verbose=1, constraint_tolerance=1e-3, square_root=true)
```
Alternatively, solver options can be set using the `set_options!` command after the
solver has been instantiated:
```julia
set_options!(solver, reset_duals=true, penalty_initial=100, penalty_scaling=50)
```

### Querying Solver Options
The options struct for the `ALTROSolver` can be directly accessed via `solver.opts` or
`Altro.options(solver)`. Options can be directly set or retrieved from this mutable
struct.

### List of Options
For convenience, we provide a list of options in the ALTRO solver, along with a brief
description:

| Option | Description | Importance | Default |
| ------ | ----------- | ---------- | ------- |
| `constraint_tolerance` | All constraint violations must be below this value. | High | `1e-6` |
| `cost_tolerance` | The difference in costs between subsequent iterations must be below this value. | High | `1e-4` |
| `cost_tolerance_intermediate` | Cost tolerance for intermediate iLQR solves. Can speed up convergence by increase to 10-100x the `cost_tolerance`. | Med | `1e-4` |
| `gradient_tolerance` | Tolerance for 2-norm of primal optimality residual. | Low | `1` |
| `gradient_tolerance_intermediate` | Primal optimality residual tolerance for intermediate solve. | Low | `10` |
| `iterations_inner` | Max iLQR iterations per iLQR solve. | Med | `300` |
| `dJ_counter_limit` | Max number of times iLQR can fail to make progress before exiting. | Low | `10` |
| `square_root` | Enable the square root backward pass for improved numerical conditioning (WIP). | Med | `false` |
| `line_search_lower_bound` | Lower bound for Armijo line search. | Low | `1e-8` |
| `line_search_upper_bound` | Upper bound for Armijo line search. | Low | `10.0` |
| `iterations_linesearch` | Max number of backtracking steps in iLQR line search | Low | 20 |
| `max_cost_value` | Maximum cost value. Will terminate solve if cost exeeds this limit. | Low | `1e8` |
| `max_state_value` | Maximum value of any state. Will terminate solve if any state exeeds this limit. | Low | `1e8` |
| `max_control_value` | Maximum value of any control. Will terminate solve if any control exeeds this limit. | Low | `1e8` |
| `static_bp` | Enable the static backward pass. Only advisable for state + control dimensions < 20. Turn off if compile time is exessive. | Low | `true` |
| `save_S` | Save the intermediate cost-to-go expansions in the iLQR backward pass. |Low | `false` |
| `bp_reg` | Enable iLQR backward pass regularization (WIP). | Med | `false` |
| `bp_reg_initial` | Initial backward pass regularization. | Low | `0.0` |
| `bp_reg_increase_factor` | Multiplicative factor by which the regularization is increased. | Low | `1.6` |
| `bp_reg_max` | Maximum regularization. | Low | `1e8` |
| `bp_reg_min` | Minimum regularization. | Low | `1e-8` |
| `bp_reg_fp` | Amount of regularization added when foward pass fails | Low | `10.0` |
| `penalty_initial` | Initial penalty term on all constraints. Set low if the unconstrained solution is a good approximate solution to the constrained problem, and high if the initial guess provided is a good esimate. If `NaN` uses values in each constraint param, which defaults to `1.0`. | `Very High` | `NaN` |
| `penalty_scaling` | Multiplicative factor by which the penalty is increased each outer loop iteration. High values can speed up convergence but quickly lead to poor numerical conditioning on difficult problems. Start with small values and then increase.If `NaN` defaults to `10` in the per-constraint parameter. | `Very High` | `NaN` |
| `iterations_outer` | Max number of outer loop (Augmented Lagrangian) iterations. | Med | `30` |
| `verbose_pn` | Turn on printing in the projected newton solver. | Low | `false` |
| `n_steps` | Maximum number of projected newton steps. | Low | `2` |
| `projected_newton_tolerance` | Constraint tolerance at which the solver will exit the Augmented Lagrangian solve and start the projected newton solve. Typically `sqrt(constraint_tolerance)` | High | `1e-3` |
| `active_set_tolerance_pn` | Tolerance for the active constraints during the projected newton solve. Includes some barely satisfied constraints into the active set. Can fix singularity issues during projected newton solve. | Med | `1e-3` |
| `multiplier_projected` | Enable updating the dual variables during the projected newton solve. Also provides a calculation of the optimality residual in the stats output. | Low | `true` |
| `ρ_chol` | Regularization on the projected newton Cholesky solve. | Med | `1e-2` |
| `ρ_primal` | Regularization on the primal variables during the projected newton solve. Required if cost Hessian is positive-semi-definite. | Low | `1e-8` |
| `ρ_dual` | Regularization on the dual variables during the multiplier projection step. | Low | `1e-8` |
| `r_threshold` | Improvement ratio threshold for projected newton solve. If the ratio of constraint violations between subsequent steps is less than this value, it will update the cost and constraint expansions | Low | `1.1` |
| `projected_newton` | Enable projected newton solve. If enabled, `projected_newton_solve` is used as the `constraint_tolerance` for the AL-iLQR solve. Projected newton solve is still a WIP and not very robust. | High | `true` |
| `iterations` | Max number of total iterations (iLQR + projected newton). | Med | 1000 |
| `verbose` | Controls output during solve. `0` is zero output, `1` outputs AL iterations, and `2` outputs both AL and iLQR iterations | Low | `0` |


NEWS.md
# `v0.5comp` Branch
This branch has both the old version of Altro from `v0.4` which uses StaticArrays and is 
very fast for small dimensions but has problems scaling up to larger dimensions,
and the new versions of the Altro solvers from `v0.5`, which uses normal Julia dynamics 
arrays, the new RobotDynamics `v0.4` API, and Octavian for performance on in-place matrix 
operations. The new solvers and methods generally have a `2` suffix, e.g. `iLQRSolver2`.
This branch should stay around for a while for the sake of comparison.

# New in `v0.2`

## Flattened API
* The new API now only exports the `ALTROSolver` type, no longer exporting `AugmentedLagrangianSolver`, `iLQRSolver`, and `ProjectedNewtonSolver`.
* Use the `projected_newton` solver option to turn off to get the same behavior as `AugmentedLagrangianSolver`. Passing an unconstrained problem is equivalent to solving with `iLQRSolver`.
* Solver options are now shared between all solvers, and implemented as `SolverOptions`. Use `solver.opts` or `set_options!(solver; opts...)` to change solver options. Solver options can also be passed into the solver constructor.
* Solver stats have also been improved, and contains a single log for all iteration types. Accessible via `solver.stats` or `Altro.stats(solver)`.

## Improved Verbosity Output
* iLQR iterations now provide a brief description for why they terminate.
* The `show_summary` option can be used to display a brief summary of the solve upon completion.

## Solver status and convergence 
* Each solve now saves a solver status in `solver.stats.status`, accessible via `status(solver)`, which provides some information about how the solver exited.
* Runtime errors in Altro.jl were removed in favor of exiting the solver with a specific solver status. Runtime errors are still possible if they originate from TrajetoryOptimization.jl or bugs (nothing is wrapped in `try/catch` loops).
* Changed convergence criteria for the iLQR solver: it must satisfy both the change in cost tolerance and the gradient tolerance at the same time.
* The `ProjectedNewtonSolver` now does a projection of the dual variables in order to accurately calculation the residual of dual optimality (gradient of the Lagrangian). Can be disabled via the `multiplier_projection` option.


src/solver_opts.jl
export
    SolverOptions


abstract type AbstractSolverOptions{T<:Real} end

"""
    set_options!(opts::AbstractSolverOptions; kwargs...)
    set_options!(solver::AbstractSolver; opts...)

Set solver options via keyword arguments, supporting either solvers or solver option
types directly. Will set any and all options that match the provided arguments. For 
example, `set_options!(solver::ALTROSolver, constraint_tolerance=1e-4)` will set the
constraint tolerance option in the ALTRO, Augmented Lagrangian, and Project Newton 
solvers.

The only exeption is the `verbose` setting, which always accepts a boolean, while
ALTRO and Augmented Lagrangian solvers accept integers 0-2, with 1 providing output
for the outer AL iterations but not the iLQR iterations.
"""
function set_options!(opts::OPT; d...) where OPT <: AbstractSolverOptions
    for (key,val) in pairs(d)
        if hasfield(OPT, key) 
            val = convert(fieldtype(OPT, key), val)
            setfield!(opts, key, val) 
        end
    end
end


Base.@kwdef mutable struct SolverOptions{T} <: AbstractSolverOptions{T}
    # Optimality Tolerances
    constraint_tolerance::T = 1e-6
    cost_tolerance::T = 1e-4
    cost_tolerance_intermediate::T = 1e-4
    gradient_tolerance::T = 10.0
    gradient_tolerance_intermediate::T = 1.0

    # iLQR
    expected_decrease_tolerance::Float64 = 1e-10
    iterations_inner::Int = 300
    dJ_counter_limit::Int = 10
    square_root::Bool = false
    line_search_lower_bound::T = 1e-8
    line_search_upper_bound::T = 10.0
    line_search_decrease_factor::T = 0.5
    iterations_linesearch::Int = 20
    max_cost_value::T = 1.0e8
    max_state_value::T = 1.0e8
    max_control_value::T = 1.0e8
    static_bp::Bool = true
	save_S::Bool = false
    closed_loop_initial_rollout::Bool = false

    # Backward pass regularization
    bp_reg::Bool = false
    bp_reg_initial::T = 0.0
    bp_reg_increase_factor::T = 1.6
    bp_reg_max::T = 1.0e8
    bp_reg_min::T = 1.0e-8
    bp_reg_type::Symbol = :control
    bp_reg_fp::T = 10.0

    # Augmented Lagrangian
    use_conic_cost::Bool = false
    penalty_initial::T = 1.0 
    penalty_scaling::T = 10.0
    penalty_max::T = 1e8 
    dual_max::T = 1e8 

    active_set_tolerance_al::T = 1e-3
    iterations_outer::Int = 30
    kickout_max_penalty::Bool = false
    reset_duals::Bool = true
    reset_penalties::Bool = true

    # Projected Newton
    force_pn::Bool = false
    verbose_pn::Bool = false
    n_steps::Int = 2
    solve_type::Symbol = :feasible
    projected_newton_tolerance::T = 1e-3
    active_set_tolerance_pn::T = 1e-3
    multiplier_projection::Bool = true
    ρ_chol::T = 1e-2     # cholesky factorization regularization
    ρ_primal::T = 1.0e-8 # primal regularization
    ρ_dual::T = 1.0e-8   # regularization for multiplier projection 
    r_threshold::T = 1.1

    # General options
    dynamics_funsig::RobotDynamics.FunctionSignature = StaticReturn()
    dynamics_diffmethod::RobotDynamics.DiffMethod = ForwardAD()
    projected_newton::Bool = true
    reuse_jacobians::Bool = false
    trim_stats::Bool = true  # disable if you want to call methods after solve that add to history
    iterations::Int = 1000   # max number of iterations
    show_summary::Bool = true 
    verbose::Int = 0 
end

function Base.copy(opts::SolverOptions)
    SolverOptions([getfield(opts, fname) for fname in fieldnames(SolverOptions)]...)
end

Base.@kwdef mutable struct SolverStats{T}
    # Iteration counts
    iterations::Int = 0
    iterations_outer::Int = 0
    iterations_pn::Int = 0

    # Iteration stats
    iteration::Vector{Int} = Int[] 
    iteration_outer::Vector{Int} = Int[]
    iteration_pn::Vector{Bool} = Bool[]   # is the iteration a pn iteration
    cost::Vector{T} = Float64[]
    dJ::Vector{T} = Float64[]
    c_max::Vector{T} = Float64[]
    gradient::Vector{T} = Float64[] 
    penalty_max::Vector{T} = Float64[]

    # iLQR
    dJ_zero_counter::Int = 0
    ls_failed::Bool = false

    # Other
    tstart::Float64 = time()
    tsolve::Float64 = Inf 
    to::TimerOutput = TimerOutput()
    status::TerminationStatus = UNSOLVED
    is_reset::Bool = false
    "Which solver is the top-level solver and responsible for resetting and trimming."
    parent::Symbol
end

function reset!(stats::SolverStats, N::Int, parent::Symbol)
    if parent == stats.parent
        stats.is_reset = false  # force a reset
        reset!(stats, N)
    end
    return nothing
end

function reset!(stats::SolverStats, N::Int=0)
    stats.is_reset && return nothing
    stats.iterations = 0
    stats.iterations_outer = 0
    stats.iterations_pn = 0
    function reset!(v::AbstractVector{T}, N::Int) where T 
        resize!(v, N)
        v .= zero(T)
    end
    reset!(stats.iteration, N)
    reset!(stats.iteration_outer, N)
    reset!(stats.iteration_pn, N)
    reset!(stats.cost, N)
    reset!(stats.dJ, N)
    reset!(stats.c_max, N)
    reset!(stats.gradient, N)
    reset!(stats.penalty_max, N)
    stats.tstart = time_ns()
    stats.tsolve = Inf
    # reset_timer!(stats.to)
    stats.dJ_zero_counter = 0
    stats.is_reset = true
    stats.status = UNSOLVED
    return nothing
end

function record_iteration!(stats::SolverStats; 
        cost=NaN, dJ=NaN, c_max=NaN, gradient=NaN, penalty_max=NaN, 
        is_pn::Bool=false, is_outer::Bool=false
    )
    # don't increment the iteration for an outer loop
    if is_outer
        stats.iterations_outer += 1
    else
        stats.iterations += 1
    end
    i = stats.iterations
    function record!(vec, val, i)
        if isnan(val)
            if vec[i] != 0
                val = vec[i]
            elseif i > 1
                val = vec[i-1]
            else
                val = vec[i]
            end
        end
        vec[i] = val
    end
    stats.iteration[i] = i
    stats.iteration_outer[i] = stats.iterations_outer
    stats.iteration_pn[i] = is_pn
    record!(stats.cost, cost, i)
    record!(stats.dJ, dJ, i)
    record!(stats.c_max, c_max, i)
    record!(stats.gradient, gradient, i)
    record!(stats.penalty_max, penalty_max, i)
    is_pn && (stats.iterations_pn += 1)
    return stats.iterations
end

function trim!(stats::SolverStats, parent::Symbol)
    if parent == stats.parent
        trim!(stats)
    end
end
function trim!(stats::SolverStats)
    N = stats.iterations
    resize!(stats.iteration, N)
    resize!(stats.iteration_outer, N)
    resize!(stats.iteration_pn, N)
    resize!(stats.cost, N)
    resize!(stats.dJ, N)
    resize!(stats.c_max, N)
    resize!(stats.gradient, N)
    resize!(stats.penalty_max, N)
    stats.is_reset = false
    return N 
end

function Dict(stats::SolverStats)
    fields = (:iteration, :iteration_outer, :iteration_pn, 
        :cost, :dJ, :c_max, :gradient, :penalty_max)
    pairs = [field=>getfield(stats,field) for field in fields]
    Dict(pairs...)
end